File: examples/solv_ore_iter.rb
#
# jruby examples for jas.
# $Id$
#

require "examples/jas"

# Ore extension solvable polynomial example, Gomez-Torrecillas, 2003

##p = PolyRing.new(QQ(),"x,y");
#is automatic: [one,x,y] = p.gens();

pcz = PolyRing.new(QQ(),"x,y,z");
#is automatic: [one,x,y,z] = p.gens();

zrelations = [z, y,  y * z + x
             ];

puts "zrelations: = [" + zrelations.join(", ") { |r| r.to_s } + "]";
puts;

pz = SolvPolyRing.new(QQ(), "x,y,z", PolyRing.lex, zrelations);
puts "SolvPolyRing: " + str(pz);
puts;

pzq = SRF(pz);
puts "SolvableQuotientRing: " + str(pzq::ring.toScript); # + ", assoz: " + str(pzq::ring.isAssociative);
#puts "gens =" + pzq::ring.generators().join(", ") { |r| r.to_s };
puts;

pct = PolyRing.new(pzq,"t");
#is automatic: [one,x,y,z,t] = p.gens();

trelations = [t, y,  y * t + y,
              t, z,  z * t - z
             ];

puts "trelations: = [" + trelations.join(", ") { |r| r.to_s } + "]";
puts;

#startLog();

pt = SolvPolyRing.new(pzq, "t", PolyRing.lex, trelations);
puts "SolvPolyRing: " + str(pt);
puts;

puts "gens =" + pt.gens().join(", ") { |r| r.to_s };
#is automatic: one,x,y,z,t = rp.gens();

#puts "t  = " + str(t);
#puts "z  = " + str(z);
zi = 1 / z;
yi = 1 / y;
xi = 1 / x;

#startLog();

a = ( t * zi ) * y;
b = t * ( zi * y );
c = a - b;
puts "t * 1/z * y: ";
puts "a   = " +str(a);
puts "b   = " +str(b);
puts "a-b = " +str(c);
puts
#bm = b.monic
#puts "monic(b) = " +str(bm);
#puts

#exit(0);

f1 = x**2 + y**2 + z**2 + t**2 + 1;
puts "f1 = " + str(f1);

f2 = t * x * f1;
puts "f2 = " + str(f2);

ff = [ f1, f2 ];
puts "ff = [" + ff.join(", ") { |r| r.to_s } + "]";
puts

ii = pt.ideal( "", ff );
puts "SolvableIdeal: " + str(ii);
puts;

#exit(0);

rgl = ii.leftGB();
puts "seq left GB: " + str(rgl);
puts "isLeftGB: " + str(rgl.isLeftGB());
puts;

#rgr = ii.rightGB();
#puts "seq right GB: " + str(rgr);
#puts "isRightGB: " + str(rgr.isRightGB());
#puts;

#startLog();

rgt = ii.twosidedGB();
puts "seq twosided GB: " + str(rgt);
puts "isTwosidedGB: " + str(rgt.isTwosidedGB());
puts;

#exit(0);
#startLog();

rgi = rgl.intersect(rgt);
puts "leftGB intersect twosidedGB: " + str(rgi);
puts;

#terminate();

File: examples/solv_ore_iter.py
#
# jython examples for jas.
# $Id$
#

from java.lang import System

from jas import SolvableRing, SolvPolyRing, PolyRing
from jas import QQ, startLog, SRC, SRF

# Ore extension solvable polynomial example, Gomez-Torrecillas, 2003

pcz = PolyRing(QQ(),"x,y,z");
#is automatic: [one,x,y,z] = pcz.gens();

zrelations = [z, y,  y * z + x
             ];

print "zrelations: = " + str([ str(f) for f in zrelations ]);
print;

pz = SolvPolyRing(QQ(), "x,y,z", PolyRing.lex, zrelations);
print "SolvPolyRing: " + str(pz);
print;

pzq = SRF(pz);
print "SolvableQuotientRing: " + str(pzq.ring.toScript()); # + ", assoz: " + str(pzq.ring.isAssociative());
#print "gens =" + str([ str(f) for f in pzq.ring.generators() ]);
print;

pct = PolyRing(pzq,"t");
#is automatic: [one,x,y,z,t] = p.gens();

trelations = [t, y,  y * t  + y,
              t, z,  z * t - z
             ];

print "relations: = " + str([ str(f) for f in trelations ]);
print;

pt = SolvPolyRing(pzq, "t", PolyRing.lex, trelations);
print "SolvPolyRing: " + str(pt);
print;
print "gens =", [ str(f) for f in pt.gens() ];

#print "t  = " + str(t);
#print "z  = " + str(z);
zi = 1 / z;
yi = 1 / y;
xi = 1 / x;

#startLog();

a = ( t * zi ) * y;
b = t * ( zi * y );
c = a - b;
print "t * 1/z * y: ";
print "a   = " +str(a);
print "b   = " +str(b);
print "a-b = " +str(c);
print
#bm = b.monic()
#print "monic(b) = " +str(bm);
#print

#exit(0);

f1 = x**2 + y**2 + z**2 + t**2 + 1;
print "f1 = ", f1;

f2 = t * x * f1;
print "f2 = ", f2;

F = [ f1, f2 ];
print "F =", [ str(f) for f in F ];
print

I = pt.ideal( list=F );
print "SolvableIdeal: " + str(I);
print;

#exit(0);

rgl = I.leftGB();
print "seq left GB:" + str(rgl);
print "isLeftGB: " + str(rgl.isLeftGB());
print;

#rgr = I.rightGB();
#print "seq right GB:" + str(rgr);
#print "isRightGB: " + str(rgr.isRightGB());
#print;

#startLog();

rgt = I.twosidedGB();
print "seq twosided GB:" + str(rgt);
print "isTwosidedGB: " + str(rgt.isTwosidedGB());
print;

#exit(0);
#startLog();

rgi = rgl.intersect(rgt);
print "leftGB intersect twosidedGB:" + str(rgi);
print;

#terminate();
